In [1]:
import pandas as pd
import numpy as np

# Load the dataset
flights = pd.read_csv('flightsFinal.csv')
flights

C:\Users\Abhishek\AppData\Local\Temp\ipykernel_18832\1095504095.py:5: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  flights = pd.read_csv('flightsFinal.csv')


YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT  SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA                    5  ...   
1                  LAX                 PBI                   10  ...   
2                  SFO                 CLT                   20  ...   
3                  LAX                 MIA                   20  ...   
4                  SEA                 ANC                   25  ...   
...                ...                 ...                  ...  ...   
5819074            LAX                 BOS                 2359  ...   
5819075            JFK                 PSE                 2359  ...   
5819076            JFK                 SJU                 2359  ...   
5819077            MCO                 SJU                 2359  ...   
5819078            JFK                 BQN                 2359  ...   

         ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED  CANCELLED  \
0               408.0          -22.0     False      False   
1               741.0           -9.0     False      False   
2               811.0            5.0     False      False   
3               756.0           -9.0     False      False   
4               259.0          -21.0     False      False   
...               ...            ...       ...        ...   
5819074         753.0          -26.0     False      False   
5819075         430.0          -16.0     False      False   
5819076         432.0           -8.0     False      False   
5819077         330.0          -10.0     False      False   
5819078         442.0            2.0     False      False   

         CANCELLATION_REASON  AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY  \
0                    UNKNOWN               0.0             0.0            0.0   
1                    UNKNOWN               0.0             0.0            0.0   
2                    UNKNOWN               0.0             0.0            0.0   
3                    UNKNOWN               0.0             0.0            0.0   
4                    UNKNOWN               0.0             0.0            0.0   
...                      ...               ...             ...            ...   
5819074              UNKNOWN               0.0             0.0            0.0   
5819075              UNKNOWN               0.0             0.0            0.0   
5819076              UNKNOWN               0.0             0.0            0.0   
5819077              UNKNOWN               0.0             0.0            0.0   
5819078              UNKNOWN               0.0             0.0            0.0   

         LATE_AIRCRAFT_DELAY  WEATHER_DELAY  
0                        0.0            0.0  
1                        0.0            0.0  
2                        0.0            0.0  
3                        0.0            0.0  
4                        0.0            0.0  
...                      ...            ...  
5819074                  0.0            0.0  
5819075                  0.0            0.0  
5819076                  0.0            0.0  
5819077                  0.0            0.0  
5819078

# Now we are changing the format like 2359 into 23:59 

In [2]:
import pandas as pd
from datetime import datetime, time

# Load the CSV file
flights = pd.read_csv("flightsFinal.csv")

# Function to convert numeric time values into proper HH:MM time format
def convert_to_time(value):
    try:
        value = int(float(value))  # Handle float and ensure it's an integer
        value_str = str(value).zfill(4)  # Ensure 4-digit format (e.g., "5" → "0005")
        hours = int(value_str[:2])  # First two digits are hours
        minutes = int(value_str[2:])  # Last two digits are minutes
        return time(hours, minutes)  # Return as datetime.time type
    except:
        return time(0, 0)  # Default to "00:00" in case of errors

# List of columns to convert
columns_to_convert = ["SCHEDULED_DEPARTURE", "DEPARTURE_TIME" , "WHEELS_ON", "WHEELS_OFF", "ARRIVAL_TIME" , "SCHEDULED_ARRIVAL"]

# Convert each column to proper time format
for col in columns_to_convert:
    if col in flights.columns:
          flights[col] = flights[col].apply(convert_to_time)



C:\Users\Abhishek\AppData\Local\Temp\ipykernel_18832\1112182115.py:5: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  flights = pd.read_csv("flightsFinal.csv")


In [3]:
flights

YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT SCHEDULED_DEPARTURE  ...  \
0                  ANC                 SEA            00:05:00  ...   
1                  LAX                 PBI            00:10:00  ...   
2                  SFO                 CLT            00:20:00  ...   
3                  LAX                 MIA            00:20:00  ...   
4                  SEA                 ANC            00:25:00  ...   
...                ...                 ...                 ...  ...   
5819074            LAX                 BOS            23:59:00  ...   
5819075            JFK                 PSE            23:59:00  ...   
5819076            JFK                 SJU            23:59:00  ...   
5819077            MCO                 SJU            23:59:00  ...   
5819078            JFK                 BQN            23:59:00  ...   

        ARRIVAL_TIME  ARRIVAL_DELAY  DIVERTED CANCELLED  CANCELLATION_REASON  \
0           04:08:00          -22.0     False     False              UNKNOWN   
1           07:41:00           -9.0     False     False              UNKNOWN   
2           08:11:00            5.0     False     False              UNKNOWN   
3           07:56:00           -9.0     False     False              UNKNOWN   
4           02:59:00          -21.0     False     False              UNKNOWN   
...              ...            ...       ...       ...                  ...   
5819074     07:53:00          -26.0     False     False              UNKNOWN   
5819075     04:30:00          -16.0     False     False              UNKNOWN   
5819076     04:32:00           -8.0     False     False              UNKNOWN   
5819077     03:30:00          -10.0     False     False              UNKNOWN   
5819078     04:42:00            2.0     False     False              UNKNOWN   

         AIR_SYSTEM_DELAY  SECURITY_DELAY  AIRLINE_DELAY LATE_AIRCRAFT_DELAY  \
0                     0.0             0.0            0.0                 0.0   
1                     0.0             0.0            0.0                 0.0   
2                     0.0             0.0            0.0                 0.0   
3                     0.0             0.0            0.0                 0.0   
4                     0.0             0.0            0.0                 0.0   
...                   ...             ...            ...                 ...   
5819074               0.0             0.0            0.0                 0.0   
5819075               0.0             0.0            0.0                 0.0   
5819076               0.0             0.0            0.0                 0.0   
5819077               0.0             0.0            0.0                 0.0   
5819078               0.0             0.0            0.0                 0.0   

         WEATHER_DELAY  
0                  0.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
...                ...  
5819074            0.0  
5819075            0.0  
5819076            0.0  
5819077            0.0  
5819078            0.0  

[5819079

In [4]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5819079 entries, 0 to 5819078
Data columns (total 31 columns):
 #   Column               Dtype  
---  ------               -----  
 0   YEAR                 int64  
 1   MONTH                int64  
 2   DAY                  int64  
 3   DAY_OF_WEEK          int64  
 4   AIRLINE              object 
 5   FLIGHT_NUMBER        int64  
 6   TAIL_NUMBER          object 
 7   ORIGIN_AIRPORT       object 
 8   DESTINATION_AIRPORT  object 
 9   SCHEDULED_DEPARTURE  object 
 10  DEPARTURE_TIME       object 
 11  DEPARTURE_DELAY      float64
 12  TAXI_OUT             float64
 13  WHEELS_OFF           object 
 14  SCHEDULED_TIME       float64
 15  ELAPSED_TIME         float64
 16  AIR_TIME             float64
 17  DISTANCE             int64  
 18  WHEELS_ON            object 
 19  TAXI_IN              float64
 20  SCHEDULED_ARRIVAL    object 
 21  ARRIVAL_TIME         object 
 22  ARRIVAL_DELAY        float64
 23  DIVERTED             bool   
 24

In [5]:
flights.isnull().sum()

YEAR                   0
MONTH                  0
DAY                    0
DAY_OF_WEEK            0
AIRLINE                0
FLIGHT_NUMBER          0
TAIL_NUMBER            0
ORIGIN_AIRPORT         0
DESTINATION_AIRPORT    0
SCHEDULED_DEPARTURE    0
DEPARTURE_TIME         0
DEPARTURE_DELAY        0
TAXI_OUT               0
WHEELS_OFF             0
SCHEDULED_TIME         0
ELAPSED_TIME           0
AIR_TIME               0
DISTANCE               0
WHEELS_ON              0
TAXI_IN                0
SCHEDULED_ARRIVAL      0
ARRIVAL_TIME           0
ARRIVAL_DELAY          0
DIVERTED               0
CANCELLED              0
CANCELLATION_REASON    0
AIR_SYSTEM_DELAY       0
SECURITY_DELAY         0
AIRLINE_DELAY          0
LATE_AIRCRAFT_DELAY    0
WEATHER_DELAY          0
dtype: int64

In [6]:
print(flights)

         YEAR  MONTH  DAY  DAY_OF_WEEK AIRLINE  FLIGHT_NUMBER TAIL_NUMBER  \
0        2015      1    1            4      AS             98      N407AS   
1        2015      1    1            4      AA           2336      N3KUAA   
2        2015      1    1            4      US            840      N171US   
3        2015      1    1            4      AA            258      N3HYAA   
4        2015      1    1            4      AS            135      N527AS   
...       ...    ...  ...          ...     ...            ...         ...   
5819074  2015     12   31            4      B6            688      N657JB   
5819075  2015     12   31            4      B6            745      N828JB   
5819076  2015     12   31            4      B6           1503      N913JB   
5819077  2015     12   31            4      B6            333      N527JB   
5819078  2015     12   31            4      B6            839      N534JB   

        ORIGIN_AIRPORT DESTINATION_AIRPORT SCHEDULED_DEPARTURE  ...  \
0   

In [7]:
flights.to_csv("flights_Final.csv", index = False)